File: mycollab-web/src/main/scala/com/esofthead/mycollab/module/project/view/ProjectPresenterDataMapper.scala
/**
 * This file is part of mycollab-web.
 *
 * mycollab-web is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-web is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-web.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.project.view

import com.esofthead.mycollab.module.project.view.bug.BugPresenter
import com.esofthead.mycollab.module.project.view.file.FilePresenter
import com.esofthead.mycollab.module.project.view.message.MessagePresenter
import com.esofthead.mycollab.module.project.view.milestone.MilestonePresenter
import com.esofthead.mycollab.module.project.view.page.PagePresenter
import com.esofthead.mycollab.module.project.view.parameters._
import com.esofthead.mycollab.module.project.view.risk.IRiskPresenter
import com.esofthead.mycollab.module.project.view.settings.UserSettingPresenter
import com.esofthead.mycollab.module.project.view.task.TaskPresenter
import com.esofthead.mycollab.module.project.view.time.IFinancePresenter
import com.esofthead.mycollab.module.project.view.user.ProjectDashboardPresenter
import com.esofthead.mycollab.vaadin.mvp.{IPresenter, ScreenData}

/**
  * @author MyCollab Ltd.
  * @since 5.0.3
  */
object ProjectPresenterDataMapper {
  val milestoneMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[MilestoneScreenData.Read] -> classOf[MilestonePresenter],
    classOf[MilestoneScreenData.Search] -> classOf[MilestonePresenter],
    classOf[MilestoneScreenData.Add] -> classOf[MilestonePresenter],
    classOf[MilestoneScreenData.Edit] -> classOf[MilestonePresenter],
    classOf[MilestoneScreenData.Roadmap] -> classOf[MilestonePresenter])

  val messageMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[MessageScreenData.Read] -> classOf[MessagePresenter],
    classOf[MessageScreenData.Search] -> classOf[MessagePresenter])

  val pageMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[PageScreenData.Read] -> classOf[PagePresenter],
    classOf[PageScreenData.Add] -> classOf[PagePresenter],
    classOf[PageScreenData.Edit] -> classOf[PagePresenter],
    classOf[PageScreenData.Search] -> classOf[PagePresenter])

  val riskMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[RiskScreenData.Add] -> classOf[IRiskPresenter],
    classOf[RiskScreenData.Edit] -> classOf[IRiskPresenter],
    classOf[RiskScreenData.Read] -> classOf[IRiskPresenter],
    classOf[RiskScreenData.Search] -> classOf[IRiskPresenter])

  val taskMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[TaskScreenData.Add] -> classOf[TaskPresenter],
    classOf[TaskScreenData.Edit] -> classOf[TaskPresenter],
    classOf[TaskScreenData.Read] -> classOf[TaskPresenter],
    classOf[TaskScreenData.GotoDashboard] -> classOf[TaskPresenter],
    classOf[TaskScreenData.GotoKanbanView] -> classOf[TaskPresenter])

  val trackerMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[BugScreenData.Read] -> classOf[BugPresenter],
    classOf[BugScreenData.Add] -> classOf[BugPresenter],
    classOf[BugScreenData.Edit] -> classOf[BugPresenter],
    classOf[BugScreenData.Search] -> classOf[BugPresenter],
    classOf[BugScreenData.GotoKanbanView] -> classOf[BugPresenter])

  val standupMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[StandupScreenData.Search] -> classOf[ProjectDashboardPresenter],
    classOf[StandupScreenData.Add] -> classOf[ProjectDashboardPresenter])

  val userMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[ProjectMemberScreenData.Add] -> classOf[UserSettingPresenter],
    classOf[ProjectMemberScreenData.InviteProjectMembers] -> classOf[UserSettingPresenter],
    classOf[ProjectMemberScreenData.Read] -> classOf[UserSettingPresenter],
    classOf[ProjectMemberScreenData.Search] -> classOf[UserSettingPresenter],
    classOf[ProjectRoleScreenData.Add] -> classOf[UserSettingPresenter],
    classOf[ProjectRoleScreenData.Read] -> classOf[UserSettingPresenter],
    classOf[ProjectRoleScreenData.Search] -> classOf[UserSettingPresenter],
    classOf[ProjectSettingScreenData.ViewSettings] -> classOf[UserSettingPresenter],
    classOf[ComponentScreenData.Add] -> classOf[UserSettingPresenter],
    classOf[ComponentScreenData.Edit] -> classOf[UserSettingPresenter],
    classOf[ComponentScreenData.Read] -> classOf[UserSettingPresenter],
    classOf[ComponentScreenData.Search] -> classOf[UserSettingPresenter],
    classOf[VersionScreenData.Add] -> classOf[UserSettingPresenter],
    classOf[VersionScreenData.Edit] -> classOf[UserSettingPresenter],
    classOf[VersionScreenData.Read] -> classOf[UserSettingPresenter],
    classOf[VersionScreenData.Search] -> classOf[UserSettingPresenter])

  val timeMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[TimeTrackingScreenData.Search] -> classOf[IFinancePresenter],
    classOf[InvoiceScreenData.GotoInvoiceList] -> classOf[IFinancePresenter])

  val fileMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[FileScreenData.GotoDashboard] -> classOf[FilePresenter])

  val projectMapper = Map[Class[_ <: ScreenData[_]], Class[_ <: IPresenter[_]]](
    classOf[ProjectScreenData.GotoTagList] -> classOf[ProjectDashboardPresenter],
    classOf[ProjectScreenData.GotoFavorite] -> classOf[ProjectDashboardPresenter],
    classOf[ProjectScreenData.GotoGanttChart] -> classOf[ProjectDashboardPresenter],
    classOf[ProjectScreenData.GotoCalendarView] -> classOf[ProjectDashboardPresenter],
    classOf[ProjectScreenData.GotoReportConsole] -> classOf[ProjectDashboardPresenter],
    classOf[ProjectScreenData.SearchItem] -> classOf[UserProjectDashboardPresenter],
    classOf[ProjectScreenData.Edit] -> classOf[UserProjectDashboardPresenter])

  val allMapper = milestoneMapper ++ messageMapper ++ pageMapper  ++ riskMapper ++ taskMapper ++
    trackerMapper ++ standupMapper ++ userMapper ++ timeMapper ++ fileMapper ++ projectMapper

  def presenter(screenData: ScreenData[_]): Class[_ <: IPresenter[_]] = {
    val _value = allMapper get screenData.getClass
    return if (_value.isInstanceOf[Some[_]]) _value.get else null
  }
}

File: mycollab-web/src/main/java/com/mycollab/module/project/view/ProjectPresenterDataMapper.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.project.view

import com.mycollab.module.project.view.bug.BugAddPresenter
import com.mycollab.module.project.view.bug.BugReadPresenter
import com.mycollab.module.project.view.finance.IInvoiceListPresenter
import com.mycollab.module.project.view.finance.ITimeTrackingPresenter
import com.mycollab.module.project.view.message.MessageListPresenter
import com.mycollab.module.project.view.message.MessageReadPresenter
import com.mycollab.module.project.view.milestone.MilestoneAddPresenter
import com.mycollab.module.project.view.milestone.MilestoneListPresenter
import com.mycollab.module.project.view.milestone.MilestoneReadPresenter
import com.mycollab.module.project.view.milestone.MilestoneRoadmapPresenter
import com.mycollab.module.project.view.page.PageAddPresenter
import com.mycollab.module.project.view.page.PageListPresenter
import com.mycollab.module.project.view.page.PageReadPresenter
import com.mycollab.module.project.view.parameters.*
import com.mycollab.module.project.view.risk.IRiskAddPresenter
import com.mycollab.module.project.view.risk.IRiskReadPresenter
import com.mycollab.module.project.view.settings.*
import com.mycollab.module.project.view.task.TaskAddPresenter
import com.mycollab.module.project.view.task.TaskReadPresenter
import com.mycollab.module.project.view.ticket.ITicketKanbanPresenter
import com.mycollab.module.project.view.ticket.TicketDashboardPresenter
import com.mycollab.module.project.view.user.ProjectDashboardPresenter
import com.mycollab.vaadin.mvp.IPresenter
import com.mycollab.vaadin.mvp.ScreenData

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
object ProjectPresenterDataMapper {
    private val milestoneMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            MilestoneScreenData.Read::class.java to MilestoneReadPresenter::class.java,
            MilestoneScreenData.Search::class.java to MilestoneListPresenter::class.java,
            MilestoneScreenData.Add::class.java to MilestoneAddPresenter::class.java,
            MilestoneScreenData.Edit::class.java to MilestoneAddPresenter::class.java,
            MilestoneScreenData.Roadmap::class.java to MilestoneRoadmapPresenter::class.java
    )

    private val messageMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            MessageScreenData.Read::class.java to MessageReadPresenter::class.java,
            MessageScreenData.Search::class.java to MessageListPresenter::class.java)

    private val pageMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            PageScreenData.Read::class.java to PageReadPresenter::class.java,
            PageScreenData.Add::class.java to PageAddPresenter::class.java,
            PageScreenData.Edit::class.java to PageAddPresenter::class.java,
            PageScreenData.Search::class.java to PageListPresenter::class.java)

    private val riskMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            RiskScreenData.Add::class.java to IRiskAddPresenter::class.java,
            RiskScreenData.Edit::class.java to IRiskAddPresenter::class.java,
            RiskScreenData.Read::class.java to IRiskReadPresenter::class.java)

    private val taskMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            TaskScreenData.Add::class.java to TaskAddPresenter::class.java,
            TaskScreenData.Edit::class.java to TaskAddPresenter::class.java,
            TaskScreenData.Read::class.java to TaskReadPresenter::class.java,
            TicketScreenData.GotoDashboard::class.java to TicketDashboardPresenter::class.java,
            TicketScreenData.GotoKanbanView::class.java to ITicketKanbanPresenter::class.java)

    private val trackerMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            BugScreenData.Read::class.java to BugReadPresenter::class.java,
            BugScreenData.Add::class.java to BugAddPresenter::class.java,
            BugScreenData.Edit::class.java to BugAddPresenter::class.java)

    private val standupMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            StandupScreenData.Search::class.java to ProjectDashboardPresenter::class.java)

    private val userMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            ProjectMemberScreenData.Add::class.java to ProjectMemberEditPresenter::class.java,
            ProjectMemberScreenData.InviteProjectMembers::class.java to ProjectMemberInvitePresenter::class.java,
            ProjectMemberScreenData.Read::class.java to ProjectMemberReadPresenter::class.java,
            ProjectMemberScreenData.Search::class.java to ProjectMemberListPresenter::class.java,
            ProjectRoleScreenData.Add::class.java to ProjectRoleAddPresenter::class.java,
            ProjectRoleScreenData.Read::class.java to ProjectRoleReadPresenter::class.java,
            ProjectRoleScreenData.Search::class.java to ProjectRoleListPresenter::class.java,
            ProjectSettingScreenData.ViewSettings::class.java to ProjectCustomPresenter::class.java,
            ComponentScreenData.Add::class.java to ComponentAddPresenter::class.java,
            ComponentScreenData.Edit::class.java to ComponentAddPresenter::class.java,
            ComponentScreenData.Read::class.java to ComponentReadPresenter::class.java,
            ComponentScreenData.Search::class.java to ComponentListPresenter::class.java,
            VersionScreenData.Add::class.java to VersionAddPresenter::class.java,
            VersionScreenData.Edit::class.java to VersionAddPresenter::class.java,
            VersionScreenData.Read::class.java to VersionReadPresenter::class.java,
            VersionScreenData.Search::class.java to VersionListPresenter::class.java)

    private val timeMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            TimeTrackingScreenData.Search::class.java to ITimeTrackingPresenter::class.java,
            InvoiceScreenData.GotoInvoiceList::class.java to IInvoiceListPresenter::class.java)

    private val projectMapper = mapOf<Class<out ScreenData<Any>>, Class<out IPresenter<*>>>(
            ProjectScreenData.GotoTagList::class.java to ProjectDashboardPresenter::class.java,
            ProjectScreenData.GotoFavorite::class.java to ProjectDashboardPresenter::class.java,
            ProjectScreenData.GotoReportConsole::class.java to ProjectDashboardPresenter::class.java,
            ProjectScreenData.Edit::class.java to ProjectDashboardPresenter::class.java,
            ReportScreenData.GotoWeeklyTiming::class.java to ProjectDashboardPresenter::class.java)

    private val allMapper = milestoneMapper + messageMapper + pageMapper + riskMapper + taskMapper +
            trackerMapper + standupMapper + userMapper + timeMapper + projectMapper

    @JvmStatic
    fun presenter(screenData: ScreenData<Any>): Class<out IPresenter<*>>? = allMapper[screenData.javaClass]
}